ownCloud/Viewer/PDFThumbnailsCollectionViewController.swift

```diff
@@ -0,0 +1,194 @@
+        return 1
+    }
+
+    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return self.pdfDocument?.pageCount ?? 0
+    }
+
+    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: PDFThumbnailCollectionViewCell.identifier, for: indexPath) as? PDFThumbnailCollectionViewCell
+        let (image, label) =  getCachedThumbnailOrFetchAsync(at: indexPath)
+        cell?.imageView?.image = image
+        cell?.pageLabel?.text = label
+        return cell!
+    }
+
+    // MARK: - UICollectionViewDelegate
+
+    override func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        if let pdfPage = self.pdfDocument?.page(at: indexPath.item) {
+            self.dismiss(animated: true) {
+                NotificationCenter.default.post(name: PDFViewerViewController.PDFGoToPageNotification.name, object: pdfPage)
+            }
+        }
+    }
+
+    override func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
+        if !decelerate {
+            updateVisibleCells()
+        }
+    }
+
+    override func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
+        updateVisibleCells()
+    }
+
+    // МАРК: - UICollectionViewDataSourcePrefetching
+
+    func collectionView(_ collectionView: UICollectionView, prefetchItemsAt indexPaths: [IndexPath]) {
+
+        for indexPath in indexPaths {
+            if let pdfPage = self.pdfDocument?.page(at: indexPath.item) {
+                if let layout = self.collectionView?.collectionViewLayout as? UICollectionViewFlowLayout {
+                    fetchThumbnail(for: pdfPage, indexPath: indexPath, size: layout.itemSize)
+                }
+            }
+        }
+    }
+
+    func collectionView(_ collectionView: UICollectionView, cancelPrefetchingForItemsAt indexPaths: [IndexPath]) {
+
+        for indexPath in indexPaths {
+            cancelFetchingThumbnail(indexPath: indexPath)
+        }
+    }
+
+    // MARK: - Private methods
+
+    fileprivate func fetchThumbnail(for page:PDFPage, indexPath:IndexPath, size:CGSize) {
+        let blockOperation = BlockOperation()
+        weak var weakBlockOperation = blockOperation
+
+        blockOperation.addExecutionBlock { [unowned self] in
+            let thumbnailImage = page.thumbnail(of: size, for: .cropBox)
+            self.thumbnailCache.setObject(thumbnailImage, forKey: page.label! as NSString)
+            if (weakBlockOperation?.isCancelled)! {
+                return
+            }
+            DispatchQueue.main.async {
+                if let cell : PDFThumbnailCollectionViewCell = self.collectionView?.cellForItem(at: indexPath) as? PDFThumbnailCollectionViewCell {
```

Comment:
I've seen this in multiple places and it's only my preference of styling things, but you can chain `if let` and avoid a los of `{ }`

Something like:
```Swift
if let cell : PDFThumbnailCollectionViewCell = self.collectionView?.cellForItem(at: indexPath) as? PDFThumbnailCollectionViewCell, 
   let visibleCells = self.collectionView?.visibleCells,
   visibleCells.contains(cell) {
                            cell.imageView?.image = thumbnailImage
                            cell.pageLabel?.text = page.label
}
```

Reply:
@pablocarmu yes, true. I can change it.

Reply:
@pablocarmu I checked once again, here chaining would sacrifice readability and in fact I only found one single other piece of code where chaining could have been applied.